### PR TITLE
Cip 12 pr

### PR DESCRIPTION
**Pull request by chiguireitor**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,17 +20,13 @@ For a simple Docker-based install of the Counterparty software stack, see [this
 
 # Manual installation
 
-Download the newest [patched Bitcoin Core](https://github.com/btcdrak/bitcoin/releases) and create
-a `bitcoin.conf` file with the following options:
+Download and install [BCoin](http://bcoin.io/) and create
+a `bcoin.conf` file with the following options:
```
Comment (deweller) [inline README.md]: Please remove all of the bcoin updates for this pull request.  We should move bcoin support to a separate PR.

File: counterpartylib/lib/backend/__init__.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from counterpartylib.lib import config
 from counterpartylib.lib import exceptions
 
-from counterpartylib.lib.backend import addrindex, btcd
+from counterpartylib.lib.backend import addrindex, btcd, bcoin
```
Comment (deweller) [inline counterpartylib/lib/backend/__init__.py]: Please move to separate pr

File: counterpartylib/lib/backend/bcoin.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+                tx_hash in noncached_txhashes, tx_hash in txhash_list, list(txhash_list.difference(noncached_txhashes)) ))
+            if  _retry < GETRAWTRANSACTION_MAX_RETRIES: #try again
+                r = getrawtransaction_batch([tx_hash], verbose=verbose, skip_missing=skip_missing, _retry=_retry+1)
+                result[tx_hash] = r[tx_hash]
+            else:
+                raise #already tried again, give up
+
+    return result
+
+# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4
```
Comment (deweller) [inline counterpartylib/lib/backend/bcoin.py]: Please move this file to another PR.

File: counterpartylib/lib/blocks.py [hunk 4]
```diff
@@ -1115,7 +1142,10 @@ def follow(db):
         # and try again repeatedly.
         try:
             block_count = backend.getblockcount()
+            logger.debug('Backend block count {}, current block {}'.format(block_count, block_index))
         except (ConnectionRefusedError, http.client.CannotSendRequest, backend.addrindex.BackendRPCError) as e:
+            logger.debug('Backedn refused connection')
```
Comment (deweller) [inline counterpartylib/lib/blocks.py]: typo on Backend

File: counterpartylib/lib/config.py [hunk 5]
```diff
@@ -110,4 +128,14 @@
...
 
+ADDRESS_OPTION_REQUIRE_MEMO = 1
+ADDRESS_OPTION_MAX_VALUE = ADDRESS_OPTION_REQUIRE_MEMO # Or list of all the address options
+
+# From what block on does CIP12 activates
+BLOCK_START_ADDRESSES_OPTIONS_MAINNET = 0
+BLOCK_START_ADDRESSES_OPTIONS_TESTNET = 0
+BLOCK_START_ADDRESSES_OPTIONS_REGTEST = 0
+
+BLOCK_START_ADDRESSES_OPTIONS = 0
```
Comment (deweller) [inline counterpartylib/lib/config.py]: This should be moved to protocol_changes.json.

File: counterpartylib/lib/messages/versions/enhanced_send.py [hunk 6]
```diff
@@ -77,6 +77,16 @@ def validate (db, source, destination, asset, quantity, memo_bytes, block_index)
     if memo_bytes is not None and len(memo_bytes) > MAX_MEMO_LENGTH:
       problems.append('memo is too long')
 
+    cursor = db.cursor()
+    try:
+        results = cursor.execute('SELECT options FROM addresses WHERE address=?', (destination,))
+        if results:
+            result = results.fetchone()
+            if result and result['options'] & config.ADDRESS_OPTION_REQUIRE_MEMO and memo_bytes is not None and (len(memo_bytes) == 0):
```
Comment (deweller) [inline counterpartylib/lib/messages/versions/enhanced_send.py]: We're sure memo_bytes will never be `None`, right?

File: counterpartylib/lib/messages/versions/send1.py [hunk 7]
```diff
@@ -51,6 +51,17 @@ def validate (db, source, destination, asset, quantity, block_index):
         if not destination:
             problems.append('destination is required')
 
+    if block_index > config.BLOCK_START_ADDRESSES_OPTIONS:
```
Comment (deweller) [inline counterpartylib/lib/messages/versions/send1.py]: Please use `util.enabled('options_require_memo')`

File: counterpartylib/lib/messages/versions/enhanced_send.py [hunk 8]
```diff
@@ -77,6 +77,16 @@ def validate (db, source, destination, asset, quantity, memo_bytes, block_index)
     if memo_bytes is not None and len(memo_bytes) > MAX_MEMO_LENGTH:
       problems.append('memo is too long')
 
+    cursor = db.cursor()
+    try:
+        results = cursor.execute('SELECT options FROM addresses WHERE address=?', (destination,))
+        if results:
```
Comment (deweller) [inline counterpartylib/lib/messages/versions/enhanced_send.py]: We should guard this with `util.enabled('options_require_memo')`, right?

File: counterpartylib/lib/messages/versions/send2.py [hunk 9]
```diff
@@ -72,6 +72,19 @@ def validate(db, source, destination, asset, quantity, block_index):
     if util.get_balance(db, source, asset) < quantity:
         raise BalanceError('balance insufficient')
 
+    if block_index > config.BLOCK_START_ADDRESSES_OPTIONS:
+        # Check destination address options
```
Comment (deweller) [inline counterpartylib/lib/messages/versions/send2.py]: Please use `util.enabled('options_require_memo')`

File: counterpartylib/test/fixtures/vectors.py [hunk 10]
```diff
@@ -965,6 +973,21 @@
...
+                {'table': 'addresses', 'values': {
+                    'options': 1,
+                    'address': ADDR[1]
+                }}
+            ]
+        }, {
+            'comment': 'test changing options to 1 on an address with LOCKED feed',
+            'in': ({'btc_amount': 1, 'block_hash': '46ac6d09237c7961199068fdd13f1508d755483e07c57a4c8f7ff18eb33a05c93ca6a86fa2e2af82fb77a5c337146bb37e279797a3d11970aec4693c46ea5a58', 'source': ADDR[4], 'destination': '', 'block_index': DP['default_block_index'], 'fee': 10000, 'supported': 1, 'block_time': 310501000, 'tx_hash': '6b4a62b80f35b0e66df4591c8a445d453d995609e2df12afe93e742bea10dd86', 'tx_index': 502,
+                    'data': b'\x00\x00\x00\x1eR\xbb3dA\x87\xd7\x84\x00\x00\x00\x00\x00\x00\x00\x00\tOPTIONS 1'},),
+            'out': None
```
Comment (deweller) [inline counterpartylib/test/fixtures/vectors.py]: This doesn't actually verify that the address table was not changed.  You may need to write a separate test for this outside of `vectors.py`.  Or figure out another way.

File: counterpartylib/protocol_changes.json [hunk 11]
```diff
@@ -128,8 +128,15 @@
     "enhanced_sends": {
         "minimum_version_major": 9,
         "minimum_version_minor": 55,
-        "minimum_version_revision": 4,
+        "minimum_version_revision": 3,
         "block_index": 999999,
         "testnet_block_index": 0
```
Comment (deweller) [inline counterpartylib/protocol_changes.json]: Please change the testnet_block_index to `1179400` for both `short_tx_type_id` and `enhanced_sends` in order to help Circle tests pass.

File: counterpartylib/lib/messages/versions/enhanced_send.py (hunk 6 above)
Comment (chiguireitor) [inline counterpartylib/lib/messages/versions/enhanced_send.py]: Just above on line 77 there's the same check, seems they are both redundant

File: counterpartylib/lib/messages/broadcast.py [hunk 12]
```diff
@@ -211,8 +211,8 @@ def parse (db, tx, message):
     if util.enabled('broadcast_invalid_check') and status != 'valid':
         return
 
-    # Options?
-    if (util.enabled('options_require_memo')) and ('locked feed' not in status) and ('options out of range' not in status):
+    # Options? if the status is invalid the previous if should have catched it
```
Comment (deweller) [inline counterpartylib/lib/messages/broadcast.py]: Got it.  Since `broadcast_invalid_check` was enabled previously.

**Comment by chiguireitor** [comment]
Polluted branch, PRing another more clean state soon
